EXIF/XMP image metadata reading

wmatous opened the pull request:
automatic title, caption, author, tags (where available)

wmatous committed:
XMP metadata reading

automatic title, caption, author, tags (where available)

wmatous committed:
BS4/multiple authors

removed BS4 import and added support for multiple authors

wmatous committed:
add libxmp

wmatous committed:
comma

wmatous committed:
python-xmp-toolkit

libxmp does not work. not sure what the package should be named in setup.py

wmatous committed:
added libxmp to travis

wmatous committed:
removed before_install for testing

wmatous committed:
replaced  libexempi

psiemens committed:
Use exempi 2

psiemens committed:
Try libexempi 32bit

psiemens committed:
Revert libexempi version

psiemens commented:
I think the problem is that the tests aren't passing, not that exempi isn't working. Running the image tests locally produces a bunch of errors for me, most of which look like this:

```bash
dispatch-admin test dispatch.tests.test_api_images
```

```
self.title = xmp.get_array_item(xmp.get_namespace_for_prefix('dc'), 'title', 1)
AttributeError: 'NoneType' object has no attribute 'get_array_item'
```

None of the test images have XMP data included, which is fine, but that's causing the `xmp` object to be `None`. That case should be handled, and it would also be a good idea to add a new test image (and accompanying test) that includes XMP data.

wmatous committed:
print statements

wmatous committed:
test case and guards

added test case for image withxmp and exif metadata. added guards to prevent images with no data or incorrect file format from being processed

wmatous committed:
added exifread to travis

wmatous committed:
try block

for XMP reading to prevent XMPError in files with no XMP data

wmatous committed:
Merge branch '774-metadata' of https://github.com/ubyssey/dispatch into 774-metadata

wmatous committed:
added none value to except

wmatous committed:
added import for XMPError to tests

wmatous committed:
moved xmperror import

wmatous committed:
indent

wmatous committed:
changed  duplicated test name

wmatous committed:
test_image changed to response

wmatous committed:
refactored tests

caption field is not given by self.client.post().

psiemens commented on dispatch/modules/content/models.py:
```diff
@@ -451,13 +459,76 @@ def save(self, **kwargs):
+
+                    # exiftags object is a dictionary indexed by numeric tags corresponding to particular fields we may be
+                    # interested in. For a full list: https://www.sno.phy.queensu.ca/~phil/exiftool/TagNames/EXIF.html
+                    self.caption = exiftags.get(270)
```
The `Image` model doesn't have a caption field. I would just leave this out and stick with `title`. That should fix your tests as well. 

Captions are only set when images are used as attachments in articles/galleries since the caption might depend on where the image is being used.

wmatous committed:
api update and single author creation

wmatous committed:
all tests passing with functionality

models.py line 561 commented. save_tags no longer clears all tags before adding by id

wmatous commented on dispatch/modules/content/models.py:
```diff
@@ -556,7 +557,9 @@ def save_thumbnail(self, image, size, name, label, file_type):
     def save_tags(self, tag_ids):
-        self.tags.clear()
+
+        # self.tags.clear()
```
@psiemens heres the thing i took out that was deleting all my tags, im not sure why my tags arent included in the validated data

wmatous commented on dispatch/api/serializers.py:
```diff
@@ -215,6 +216,7 @@ def update(self, instance, validated_data):
 
         tag_ids = validated_data.get('tag_ids', False)
+
         if tag_ids != False:
```
@psiemens heres the bit that calls save tags on "validated data"

psiemens committed:
Remove old dependency, move code to separate method

psiemens committed:
Make author tests more robust

psiemens committed:
Fix broken EXIF/XMP tests.

psiemens committed:
Diagnose issue with authors

psiemens committed:
Diagnose issue with authors

psiemens committed:
Diagnose issue with authors

psiemens committed:
Diagnose issue with authors

psiemens committed:
Diagnose issue with authors

psiemens committed:
Diagnose issue with authors

psiemens committed:
Diagnose issue with authors

psiemens committed:
Diagnose issue with authors

psiemens committed:
Diagnose issue with authors

psiemens committed:
Diagnose issue with authors

psiemens committed:
Diagnose issue with authors

psiemens committed:
Diagnose issue with authors

psiemens committed:
Diagnose issue with authors

psiemens committed:
Move XMP parsing to separate module

psiemens committed:
Merge branch 'develop' into 774-metadata

psiemens committed:
Fix migrations

psiemens committed:
remove extra code

psiemens committed:
Clean up tests

psiemens committed:
Move all image metadata functions to separate file

psiemens committed:
Fix mistake: authors local variable
